### PR TITLE
feat: Assistant stream mode and event callback

### PR DESCRIPTION
**Pull request by CallOrRet**
This update enhances the existing functionality by adding support for Assistant stream mode and stream event callbacks without introducing any breaking changes or modifications to the existing API endpoints.

The implemented features are in line with the OpenAI Assistant streaming API, as documented here: [OpenAI Assistant Streaming API](https://platform.openai.com/docs/api-reference/assistants-streaming)

```
Stream.On("event1", callback1)
Stream.On("event2", callback2)
Stream.Run()
```

### CODE REVIEW

**Comment by sashabaranov** [comment]
Thank you for the PR! Sorry for a long review process :D

File: assistant_stream.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+var (
+	ErrStreamEventEmptyTopic    = errors.New("stream event gets a empty topic")
```
Comment (sashabaranov) [inline assistant_stream.go:11]:
```suggestion
	ErrStreamEventEmptyTopic    = errors.New("stream event gets an empty topic")
```

File: stream_reader.go [hunk 2]
```diff
@@ -28,9 +30,56 @@ type streamReader[T streamable] struct {
...
+	stream.handlers[topic] = handler
+	return nil
+}
+
+func (stream *streamReader[T]) Run() error {
+	if stream.handlerCtx == nil {
+		stream.event = "message" // default event for chat completion stream
+		ctx, cancel := context.WithCancel(context.Background())
+		stream.handlerCtx = ctx
+		go func() {
```
Comment (sashabaranov) [inline stream_reader.go:60]: Sorry, I don't think we're going to merge this as starting goroutines don't fit into the design of this library.

Please refer to the following comments:
- https://github.com/sashabaranov/go-openai/pull/54#discussion_r1083519779
- https://github.com/sashabaranov/go-openai/issues/49#issuecomment-1373423407

Comment (sashabaranov) [reply stream_reader.go:60]: I'm happy to have this in the `examples/` folder, though!

Comment (CallOrRet) [reply stream_reader.go:60]: No problem. What if I revise the parts where I used goroutines? Would that work better with the design of the library?

Comment (sashabaranov) [reply stream_reader.go:60]: @CallOrRet I'm totally open to that!

File: assistant_stream.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+var (
+	ErrStreamEventEmptyTopic    = errors.New("stream event gets an empty topic")
+	ErrStreamEventCallbackPanic = errors.New("occurs a runtime panic during the stream event callback")
+)
+
+type AssistantThreadRunStreamResponse struct {
+	ID     string       `json:"id"`
+	Object string       `json:"object"`
+	Delta  MessageDelta `json:"delta,omitempty"`
+}
```
Comment (sauravexodus) [inline assistant_stream.go:19]: This is a very trimmed down version of the response model.

File: assistant_stream.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+	resp, err := sendRequestStream[AssistantThreadRunStreamResponse](c, req)
+	if err != nil {
+		return nil, err
+	}
+
+	stream = &AssistantThreadRunStream{
+		streamReader: resp,
+	}
+	return
```
Comment (sauravexodus) [inline assistant_stream.go:51]:
```suggestion
	return stream, nil
```

File: assistant_stream.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	urlSuffix := fmt.Sprintf("/threads/%s/runs/%s/submit_tool_outputs", threadID, runID)
+	req, err := c.newRequest(
+		ctx,
+		http.MethodPost,
+		c.fullURL(urlSuffix),
+		withBody(request),
+		withBetaAssistantVersion(c.config.AssistantVersion),
+	)
+	if err != nil {
+		return
```
Comment (sauravexodus) [inline assistant_stream.go:70]:
```suggestion
		return nil, err
```

File: assistant_stream.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+	resp, err := sendRequestStream[AssistantThreadRunStreamResponse](c, req)
+	if err != nil {
+		return nil, err
+	}
+
+	stream = &AssistantThreadRunStream{
+		streamReader: resp,
+	}
+	return
```
Comment (sauravexodus) [inline assistant_stream.go:81]:
```suggestion
	return stream, nil
```

**Review by sauravexodus — COMMENTED** [review]
There are some commented code in this PR as well. But the bigger concern is the incomplete response model.

File: assistant_stream.go (hunk 3 above)
Comment (sauravexodus) [reply assistant_stream.go:19]: The complete response has a lot many parameters which would be important to run function calling etc. https://platform.openai.com/docs/api-reference/runs/createRun

File: run.go [hunk 7]
```diff
@@ -137,6 +138,7 @@ type RunList struct {
 
 type SubmitToolOutputsRequest struct {
 	ToolOutputs []ToolOutput `json:"tool_outputs"`
+	Stream      bool
```
Comment (sauravexodus) [inline run.go:141]:
```suggestion
	Stream      bool              `json:"stream,omitempty"`
```

**Pull request closed without merging**